src/analyses/threadJoins.ml

```diff
@@ -45,6 +45,19 @@ struct
         | _ -> ctx.local (* if multiple possible thread ids are joined, none of them is must joined*)
         (* Possible improvement: Do the intersection first, things that are must joined in all possibly joined threads are must-joined *)
       )
+    | `Unknown "__goblint_assume_join" ->
+      let id = List.hd arglist in
+      let threads = ctx.ask (Queries.EvalThread id) in
+      if TIDs.is_top threads then
+        D.bot () (* consider everything joined, D is reversed so bot is All threads *)
+      else (
+        (* elements throws if the thread set is top *)
+        let threads = TIDs.elements threads in
+        List.fold_left (fun acc tid ->
+            let joined = ctx.global tid in
+            D.union (D.add tid acc) joined
+          ) ctx.local threads
+      )
```

Comment:
I have a few questions here:

### Apron Privatizations

I think this does currently conflict with the apron privatizations that do not read from must-joined threads (`ana.apron.priv.must-joined`).

We currently do not join these values in the local per-mutex information, but would exclude them in the privatization then. Or am I missing something?

https://github.com/goblint/analyzer/blob/35f5b9e9ce00222e2221605df4262fe3ab6158a8/src/analyses/apron/apronPriv.apron.ml#L982-L998

Also how would it work with unknown thread ids? We would if this privatization is active have to read from all threads created anywhere?

### Threads created after call to `__goblint_assume_join(...)`

What happens if a thread with the same non-unique thread-id is created again after `__goblint_assume_join(...)`? We would need to remove it from must-joined, right?

What happens if a new thread is created after we called `__goblint_assume_join(...)` on something that was the top thread id. We would (unsoundly) already consider this thread joined?

**I think we should re-consider the handling of top here, think about the interaction with apron privatizations and definitely write tests for these things before merging.**

Reply:
It's an annotation coming from the user, so if the user makes incorrect assumptions, then the analysis is unsound, yes. We cannot check whether the assumption is always correct (if we could, then we would have a powerful enough analysis to make such annotation by the user redundant).
The point of the annotation is to be placed after a threads joining loop, which in reality ensures that all the threads have been joined. If they haven't, then that's on the user for incorrectly using the annotation.
It covers the extremely common case, where we get tons of spurious races with some cleanup code.

Regarding the top, I don't see the problem because prior to this the joined thread set had a lifted top with the literal meaning "All threads". So if the joined thread set is that top, it means _all_ threads (that could possibly be created or exist) are joined. If there's a problem with that "All threads" and it shouldn't be used, then the safest thing would be to remove that top lifting and make such situation unrepresentable.

I also don't understand how it would cause a problem for the apron analysis, because if it's not handled there, then the apron analysis continues without having joined anything, just like it did before, still accounting for the possibility of values from the threads. That would be an overapproximation, no?
I could also crudely add the handling of the annotation to the apron analysis and simply make it fail, if that seems safer.

Reply:
> The point of the annotation is to be placed after a threads joining loop, which in reality ensures that all the threads have been joined. If they haven't, then that's on the user for incorrectly using the annotation.

Fair enough. However, it seems very counterintuitive that this annotation would extend to threads that are created **AFTER** the annotation and just happen to share abstract thread ids. I think that makes it very hard to reason about as a user, who would place it at a program point to tell Goblint that all threads created thus far are joined, but would not expect this to extend to future copies of the thread being created.

As an extreme case, consider a function that spawns `n` threads, does some work and the  joins all `n` threads again. If we call this function twice (with joined in contexts), in the second call we would assume that at the second invocation all threads it creates are already joined. I don't think that this is particularly helpful.

> I also don't understand how it would cause a problem for the apron analysis, because if it's not handled there, then the apron analysis continues without having joined anything, just like it did before, still accounting for the possibility of values from the threads. 

No, it is added to the set of must joined thread ids and the apron privatization queries this set when locking mutexes no longer reading from the ones associated with must-joined threads. This only works  under the assumption that this must-joined set only contains things where all possible values associated with mutexes and must-joined threads were joined into the thread-local information the apron privatization keeps at the threadjoin, which does no longer happen here.


> If there's a problem with that "All threads" and it shouldn't be used, then the safest thing would be to remove that top lifting and make such situation unrepresentable.

There is. I think the only reason this was not done is for compatibility with the old thread ids.

Reply:
I added an example to illustrate the issue, I hope that helps more than the theoretical discussion above.

Reply:
Making threadJoins analysis remove things on `threadspawn` should be a simple change that avoids the main confusion about it affecting future thread creations. And it wouldn't hurt the annotation's intended use case, where the thread isn't recreated anyway.

I'm now quite surprised that threadJoins doesn't already have that logic though. And it's not obvious why it wouldn't be a problem for normal joins, considering we have four different thread ID domains available (plain vs history, with or without `include-node`). Is it assuming something about the thread ID domain that makes this impossible with normal thread joins?
For example, if a loop creates a single thread and joins it in the same iteration, then it could also be considered unique (only a single copy of it exists at a time) and thus must-joined. If I remember correctly, the history thread ID domain doesn't consider that unique, but other notions of thread uniqueness (like the one provided by the old thread uniqueness analysis) do AFAIK.

> I added an example to illustrate the issue, I hope that helps more than the theoretical discussion above.

Thanks! I will have a look at it.



> There is. I think the only reason this was not done is for compatibility with the old thread ids.

As for the "All threads" must-joined set, it might also be avoidable by slightly weakening what I do right now for the annotation: if the thread ID set for the argument contains an unknown thread, then just exclude that and assume-join all the known ones. Hopefully this is sufficient for the chrony case.
What this assume-joining doesn't change though, is that the base analysis still invalidates all the stuff from a `free`, even if the access analysis makes all the exemptions for ignoring those. But that means the invalidation puts unknown pointers and threads all over the flow-insensitive invariant of the global, leaking unknowns everywhere from the beginning, killing any precision we might have had there.

Reply:
I also changed the assume join behavior with unknown thread IDs now to not use "All threads" for must-joined, hopefully this gives less surprising behavior.